src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs

```diff
@@ -0,0 +1,146 @@
+using Raven.Server.Config;
+using Raven.Server.Documents.Replication;
+using Raven.Server.Documents.Replication.Incoming;
+using Raven.Server.Documents.Replication.ReplicationItems;
+using Raven.Server.Documents.TcpHandlers;
+using Raven.Server.ServerWide.Context;
+using Sparrow.Json;
+using Sparrow.Server;
+using Voron;
+
+namespace Raven.Server.Documents.Sharding.Handlers
+{
+    public class ShardedIncomingReplicationHandler : AbstractIncomingReplicationHandler<TransactionContextPool, TransactionOperationContext>
+    {
+        private readonly ShardedDatabaseContext.ShardedReplicationContext _parent;
+        private readonly ReplicationQueue _replicationQueue;
+
+        public ShardedIncomingReplicationHandler(TcpConnectionOptions tcpConnectionOptions, ShardedDatabaseContext.ShardedReplicationContext parent,
+            JsonOperationContext.MemoryBuffer buffer, ReplicationLatestEtagRequest replicatedLastEtag, ReplicationQueue replicationQueue)
+            : base(tcpConnectionOptions, buffer, parent.Server, parent.DatabaseName, replicatedLastEtag, parent.Context.DatabaseShutdown, parent.Server.ContextPool)
+        {
+            _parent = parent;
+            _replicationQueue = replicationQueue;
+            _attachmentStreamsTempFile = new StreamsTempFile("ShardedReplication" + Guid.NewGuid(), false);
+        }
+
+        protected override void EnsureNotDeleted(string nodeTag) => _parent.EnsureNotDeleted(_parent.Server.NodeTag);
+
+        protected override void OnFailed(Exception exception)
+        {
+        }
+
+        protected override void HandleHeartbeatMessage(TransactionOperationContext jsonOperationContext, BlittableJsonReaderObject blittableJsonReaderObject)
+        {
+        }
+
+        protected override void OnDocumentsReceived()
+        {
+        }
+
+        protected override void OnAttachmentStreamsReceives(int attachmentStreamCount)
+        {
+        }
+
+        protected override void HandleTaskCompleteIfNeeded()
+        {
+            _replicationQueue.SendToShardCompletion = new AsyncCountdownEvent(_parent.Context.ShardCount);
+        }
+
+        protected override Task HandleBatch(TransactionOperationContext context, IncomingReplicationHandler.DataForReplicationCommand dataForReplicationCommand, long lastEtag)
+        {
+            PrepareReplicationDataForShards(context, dataForReplicationCommand);
+            return _replicationQueue.SendToShardCompletion.WaitAsync();
+        }
+
+        private void PrepareReplicationDataForShards(TransactionOperationContext context, IncomingReplicationHandler.DataForReplicationCommand dataForReplicationCommand)
+        {
+            var dictionary = new Dictionary<int, List<ReplicationBatchItem>>();
+            for (var shard = 0; shard < _parent.Context.ShardCount; shard++)
+                dictionary[shard] = new List<ReplicationBatchItem>();
+
+            foreach (var item in dataForReplicationCommand.ReplicatedItems)
```

Comment:
DevHelper. Optimization possibility: instead of iterating over materialized batch, we can do it while reading from the stream

Reply:
done